### PR TITLE
chore(roxctl): roxctl guidelines

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR is a first step in introducing a few guidelines / common knowledge about structuring commands, writing commands, as well as maintaining commands.

While this currently doesn't cover _everything_, it serves as a starting point for additional things being covered in the future.

## Checklist
~~- [ ] Investigated and inspected CI test results~~
~~- [ ] Unit test and regression tests added~~
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed
- N/A.

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* master
  * **PR #7554** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/7554" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/7554?utm_source=stack-comment).

**Review by rhybrillou — APPROVED** [review]

**Review by janisz — APPROVED** [review]
It will be great if we could enforce all of the **must**. So far we only force environment with https://github.com/stackrox/stackrox/blob/9eecfee1f2b98327e02e72d08b44d33545af5107/.golangci.yml#L114-L119 by blocking usage of other output functions.

File: roxctl/README.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+func OldCommand(cliEnvironment environment.Environment) *cobra.Command {
+	sampleCmd := &sampleCmd{env: cliEnvironment}
+
+	cmd := &cobra.Command{
+		Use: "old-sample",
+		Short: "A sample command",
+		Long: `A sample command.
+This is a sample for a multi-line detailed description for the sample command`,
+        Hidden: true,
```
Comment (rukletsov) [inline roxctl/README.md:133]:
```suggestion
		Hidden: true,
```

File: roxctl/README.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+		Args: cobra.NoArgs,
+		// Each command **must** always use `RunE` and return an error.
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cliEnvironment.Logger().WarnLn("old-sample is deprecated. Please use new-sample.")
+			return sampleCmd.RunE(cmd, args)
+		},
+	}
+
+	cmd.Flags().StringVarP(&sampleCmd.name, "name", "n", "", "name to say hello to")
+    return cmd
```
Comment (rukletsov) [inline roxctl/README.md:143]:
```suggestion
	return cmd
```

File: roxctl/README.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+```go
+package sample
+
+func OldCommand(cliEnvironment environment.Environment) *cobra.Command {
+	sampleCmd := &sampleCmd{env: cliEnvironment}
+
+	cmd := &cobra.Command{
+		Use: "old-sample",
+		Short: "A sample command",
+		Long: `A sample command.
```
Comment (rukletsov) [inline roxctl/README.md:131]: Shall descriptions be updated to mention that the old command is deprecated and what to use instead?

File: roxctl/README.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+// require (e.g. a gRPC client) may be constructed here as well.
+func (s *sampleCmd) Construct(cmd *cobra.Command) error {
+	return nil
+}
+
+// Sample is an example for any possible business logic your command may execute.
+func (s *sampleCmd) sample() error {
+	s.env.Logger().InfofLn("Hello world, %s!", s.name)
+	return nil
+}
```
Comment (rukletsov) [inline roxctl/README.md:188]: I would ditch everything that is not related to renaming, so that readers don't need to figure what's boilerplate and what's relevant. You can keep function definitions to demonstrate that they are shared but maybe in some compact way like:
```
// Aliases can share code and reuse functions, e.g.:
func (s *sampleCmd) RunE(cmd *cobra.Command, args []string) error { ... }
func (s *sampleCmd) Construct(cmd *cobra.Command) error { ... }
func (s *sampleCmd) sample() error {. .. }
```

File: roxctl/README.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+```
```
Comment (rukletsov) [inline roxctl/README.md:189]: Empty newline at the end of the file, please.

File: roxctl/README.md (hunk 3 above)
Comment (dhaus67) [reply roxctl/README.md:144]: Added this to the `Long` description, with `Hidden` set to true, you'll only ever see the command's help if you do `roxctl old-sample --help`.

File: roxctl/README.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+// Sample command holds the cli environment, as well as all resolved flag values. This makes unit testing easier.
+type sampleCmd struct {
+	env environment.Environment
+
+	name string
+}
+
+// Construct ensure all flag values are resolved and possibly validated. Any additional external dependency that you may
+// require (e.g. a gRPC client) may be constructed here as well.
+func (s *sampleCmd) Construct(cmd *cobra.Command) error {
```
[the remark this answers is not in the document]
Comment (dhaus67) [reply roxctl/README.md:77]: When adding the environment we thought about it, but didn't do it back then.

It might be worthwhile to require this as an interface in the future, we can enhance it afterwards. I'd like to make the code changes before adding it to the README though.

**Review by rukletsov — APPROVED** [review]

File: roxctl/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rukletsov) [reply roxctl/README.md:77]: Sounds good, thank you!
